src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxTransformNode.cs

```diff
@@ -0,0 +1,118 @@
+using System.Threading;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISyntaxTransformNode
+    {
+        ISyntaxTransformBuilder GetBuilder(DriverStateTable previousStateTable);
+    }
+
+    internal interface ISyntaxTransformBuilder
+    {
+        bool Filter(SyntaxNode node);
+
+        void AddFilterFromPreviousTable(SemanticModel model, EntryState state);
+
+        void AddFilterEntries(ImmutableArray<SyntaxNode> nodes, SemanticModel model);
+
+        void SetInputState(DriverStateTable.Builder builder);
+    }
+
+    internal sealed class SyntaxTransformNode<T> : IIncrementalGeneratorNode<T>, ISyntaxTransformNode
+    {
+        private readonly Func<GeneratorSyntaxContext, T> _func;
+        private readonly IEqualityComparer<T> _comparer;
+        private readonly Func<SyntaxNode, bool> _filterFunc;
+
+        private readonly InputNode<SyntaxNode> _filterNode = new InputNode<SyntaxNode>();
+
+        internal SyntaxTransformNode(Func<SyntaxNode, bool> filterFunc, Func<GeneratorSyntaxContext, T> transformFunc, IEqualityComparer<T>? comparer = null)
+        {
+            _func = transformFunc;
+            _filterFunc = filterFunc;
+            _comparer = comparer ?? EqualityComparer<T>.Default;
+        }
+
+        // this is an input node, so we don't perform any updates
+        public NodeStateTable<T> UpdateStateTable(DriverStateTable.Builder graphState, NodeStateTable<T> previousTable, CancellationToken cancellationToken) => previousTable;
+
+        public IIncrementalGeneratorNode<T> WithComparer(IEqualityComparer<T> comparer) => new SyntaxTransformNode<T>(_filterFunc, _func, comparer);
+
+        public ISyntaxTransformBuilder GetBuilder(DriverStateTable previousStateTable) => new Builder(this, previousStateTable.GetStateTable(this), previousStateTable.GetStateTable(this._filterNode));
+
+        private sealed class Builder : ISyntaxTransformBuilder
+        {
+            private readonly SyntaxTransformNode<T> _owner;
+
+            private readonly NodeStateTable<T> _previousTransformTable;
+            private readonly NodeStateTable<T>.Builder _transformTable;
+
+            private readonly NodeStateTable<SyntaxNode> _previousFilterTable;
+            private readonly NodeStateTable<SyntaxNode>.Builder _filterTable;
+
+            public Builder(SyntaxTransformNode<T> owner, NodeStateTable<T> previousTransform, NodeStateTable<SyntaxNode> previousFilter)
+            {
+                _owner = owner;
+
+                _previousFilterTable = previousFilter;
+                _filterTable = new NodeStateTable<SyntaxNode>.Builder();
+
+                _previousTransformTable = previousTransform;
+                _transformTable = new NodeStateTable<T>.Builder();
+            }
+
+            public bool Filter(SyntaxNode syntaxNode) => _owner._filterFunc(syntaxNode);
+
+            public void AddFilterFromPreviousTable(SemanticModel model, EntryState state)
+            {
```

Comment:
```suggestion
            public void AddFilterFromPreviousTable(SemanticModel? model, EntryState state)
            {
                Debug.Assert(model is object || state == EntryState.Removed);
```

This will allow you to remove the `null!` in the driver code